### ISSUE TITLE
Inconsistent SoX speed behaviour compared to WavAugment

### ISSUE THREAD

**Issue by pzelasko** [pytorch/audio#1019]
## 🐛 Bug

With the following effect chain:
```python
def speed(sampling_rate: int) -> List[List[str]]:
    return [
        # Random speed perturbation factor between 0.9x and 1.1x the original speed
        ['speed', RandomValue(0.9, 1.1)],
        ['rate', sampling_rate],  # Resample back to the original sampling rate (speed changes it)
    ]
```

being applied to a tensor of 16000 samples (with a sampling rate of 16000) I am receiving an output that has a different number of samples than the input. This behavior is inconsistent with what WavAugment does - the shapes are always identical. 

This is tested in Lhotse [here for WavAugment](https://github.com/lhotse-speech/lhotse/pull/124/files#diff-17e8316be9b40589dfb1b8d00e8e5381db8fa807588dd5d221fa1bb0cb8e06e6R16) and [here for torchaudio](https://github.com/lhotse-speech/lhotse/pull/124/files#diff-69be996356cf07a04cce621d882ca41d31b2047ea48826fbe584b1de857122c0R16) in PR https://github.com/lhotse-speech/lhotse/pull/124.

## To Reproduce

Run Lhotse's test suite with torchaudio 0.7 installed. See e.g. the error in Lhotse's CI here: https://github.com/lhotse-speech/lhotse/pull/124/checks?check_run_id=1386141728 

## Expected behavior

Equal input and output tensor shapes.

## Environment

*This is my local MacOS env; the other one is GitHub Actions CI in Lhotse.*

PyTorch version: 1.7.0
Is debug build: True
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Mac OSX 10.15.7 (x86_64)
GCC version: Could not collect
Clang version: 12.0.0 (clang-1200.0.32.21)
CMake version: version 3.18.4

Python version: 3.7 (64-bit runtime)
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.18.1
[pip3] torch==1.7.0
[pip3] torchaudio==0.7.0a0+ac17b64
[conda] blas                      1.0                         mkl  
[conda] mkl                       2019.4                      233  
[conda] mkl-service               2.3.0            py37hfbe908c_0  
[conda] mkl_fft                   1.2.0            py37hc64f4ea_0  
[conda] mkl_random                1.1.1            py37h959d312_0  
[conda] numpy                     1.18.1           py37h7241aed_0  
[conda] numpy-base                1.18.1           py37h3304bdc_1  
[conda] pytorch                   1.7.0                   py3.7_0    pytorch
[conda] torchaudio                0.5.1                    pypi_0    pypi

_(ignore that last one version, it's some conda+pip quirk... the torchaudio version is 0.7.0 😅)_

**Comment by mthrok** [comment]
Hi @pzelasko 

How off are they each other?

My initial thought is that this is not a bug but the way WavAugment handles speed is different. 
Second thought is that based on the `libsox` being used, it generates a slightly different results. I noticed this when I was writing tests for the sox effects. (sox command installed via OS utility vs the one cmake builds as a part of torchaudio build process)

`apply_effects_tensor` and `apply_effects_file` are tested and confirmed that they return an identical result to `sox` command for `speed` as the following.

https://github.com/pytorch/audio/blob/0e1d814b76725f50b76d938b474b616a15009b45/test/torchaudio_unittest/assets/sox_effect_test_args.json#L73-L74

https://github.com/pytorch/audio/blob/0e1d814b76725f50b76d938b474b616a15009b45/test/torchaudio_unittest/sox_effect/sox_effect_test.py#L54-L79

https://github.com/pytorch/audio/blob/0e1d814b76725f50b76d938b474b616a15009b45/test/torchaudio_unittest/sox_effect/sox_effect_test.py#L103-L128

**Comment by pzelasko** [comment]
Yeah, bug might not be the best description - sorry. 

I don't think it's actually the `speed` behaviour, seems like the difference is in how the `rate` effect is being applied. Consider the following:

```
import torch
from torchaudio import sox_effects as se
audio = torch.sin(2 * 3.14 * torch.linspace(0, 1, 16000)).unsqueeze(0)

se.apply_effects_tensor(audio, 16000, [['speed', '0.9']])
Out[10]: (tensor([[ 0.0000,  0.0004,  0.0008,  ..., -0.0040, -0.0036, -0.0032]]), 14400)

se.apply_effects_tensor(audio, 16000, [['speed', '0.9']])[0].shape
Out[11]: torch.Size([1, 16000])

se.apply_effects_tensor(audio, 16000, [['speed', '0.9'], ['rate', '16000']])[0].shape
Out[12]: torch.Size([1, 17778])

se.apply_effects_tensor(audio, 16000, [['speed', '0.9'], ['rate', '-q', '16000']])[0].shape
Out[13]: torch.Size([1, 17778])
```

That last effect chain in WavAugment would have returned a tensor of shape [1, 16000]. Now that I think of it, it's reasonable that the number of samples would have changed... I'll need to check what are they doing (maybe truncating/padding the output?), if you happen to know please share :)

**Comment by pzelasko** [comment]
OK I think I understand it now. When the output length description is specified in their effect chain invocation, they are truncating or zero-padding the signal (I see that here: https://github.com/facebookresearch/WavAugment/blob/master/augment/speech_augment.h#L120). Sorry for the false alarm! 😉

**Comment by mthrok** [comment]
Glad you figured. :)
